libyul/YulControlFlowGraphExporter.cpp

```diff
@@ -0,0 +1,164 @@
+#include <libyul/backends/evm/StackHelpers.h>
+
+#include <libsolutil/Algorithms.h>
+
+#include <range/v3/view/transform.hpp>
+#include <range/v3/view/map.hpp>
+
+using namespace solidity;
+using namespace solidity::langutil;
+using namespace solidity::util;
+using namespace solidity::yul;
+
+Json YulControlFlowGraphExporter::operator()(CFG const& _cfg)
+{
+	Json ret = Json::array();
+	util::BreadthFirstSearch<CFG::BasicBlock const*> bfs{{_cfg.entry}};
+	for (auto const& functionInfo: _cfg.functionInfo | ranges::views::values)
+		bfs.verticesToTraverse.emplace_back(functionInfo.entry);
+
+	bfs.run([&](CFG::BasicBlock const* _block, auto _addChild) {
+		// Convert current block to JSON
+		ret.push_back(toJson(*_block));
+
+		Json exitBlockJson;
+		exitBlockJson["id"] = "Block" + std::to_string(getBlockId(*_block)) + "Exit";
+		exitBlockJson["instructions"] = Json::array();
+		std::visit(util::GenericVisitor{
+			[&](CFG::BasicBlock::MainExit const&) {
+				exitBlockJson["exit"] = { "Block" + std::to_string(getBlockId(*_block)) };
+				exitBlockJson["type"] = "MainExit";
+			},
+			[&](CFG::BasicBlock::Jump const& _jump)
+			{
+				exitBlockJson["exit"] = { "Block" + std::to_string(getBlockId(*_jump.target)) };
+				exitBlockJson["type"] = "Jump";
+
+				//TODO: handle backwards jump?
+				_addChild(_jump.target);
+			},
+			[&](CFG::BasicBlock::ConditionalJump const& _conditionalJump)
+			{
+				exitBlockJson["exit"] = { "Block" + std::to_string(getBlockId(*_conditionalJump.zero)), "Block" + std::to_string(getBlockId(*_conditionalJump.nonZero)) };
+				exitBlockJson["cond"] = stackSlotToJson(_conditionalJump.condition);
+				exitBlockJson["type"] = "ConditionalJump";
+
+				_addChild(_conditionalJump.zero);
+				_addChild(_conditionalJump.nonZero);
+			},
+			[&](CFG::BasicBlock::FunctionReturn const& _return) {
+				exitBlockJson["instructions"].push_back(_return.info->function.name.str());
+				exitBlockJson["exit"] = { "Block" + std::to_string(getBlockId(*_block)) };
+				exitBlockJson["type"] = "FunctionReturn";
+			},
+			[&](CFG::BasicBlock::Terminated const&) {
+				exitBlockJson["exit"] = { "Block" + std::to_string(getBlockId(*_block)) };
+				exitBlockJson["type"] = "Terminated";
+			},
+		}, _block->exit);
+		ret.push_back(exitBlockJson);
+	});
+	return ret;
+}
+
+Json YulControlFlowGraphExporter::toJson(CFG::BasicBlock const& _block)
+{
+	Json blockJson;
+	blockJson["id"] = "Block" + std::to_string(getBlockId(_block));
+	blockJson["instructions"] = Json::array();
+	for (auto const& operation: _block.operations)
+		blockJson["instructions"].push_back(toJson(operation));
+	blockJson["exit"] = "Block" + std::to_string(getBlockId(_block)) + "Exit";
+	blockJson["type"] = "BasicBlock";
+
+	return blockJson;
+}
+
+Json YulControlFlowGraphExporter::toJson(CFG::Operation const& _operation)
+{
+	Json opJson;
+	std::visit(util::GenericVisitor{
+		[&](CFG::FunctionCall const& _call) {
+			opJson["op"] = _call.function.get().name.str();
```

Comment:
assert that ``!input.empty()``, and that ``input.front()`` is a ``FunctionReturn`` slot referrring to the called function

and remove it.

Reply:
This does not seems to hold if the function call is a `panic_error`: https://github.com/ethereum/solidity/blob/c967d468fd1d6a2809ce048c978e08ecae3f16b7/libsolidity/codegen/YulUtilFunctions.cpp#L4553